pkg/main/main.go

```diff
@@ -0,0 +1,121 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strconv"
+	"strings"
+)
+
+const serversEnvVar = "SERVERS"
+const serverCmdEnvVar = "SERVER_CMD"
+
+// Constants indicating the types of errors for Dispatch function invocation
+const (
+	InputError    = "InputError"
+	FunctionError = "FunctionError"
+	SystemError   = "SystemError"
+)
+
+var servers []Server
+var router Router
+
+func handler(w http.ResponseWriter, r *http.Request) {
+	reader := r.Body
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(reader)
+	rawBody := buf.Bytes()
+	var body map[string]interface{}
+	err := json.Unmarshal(rawBody, &body)
+	if err != nil {
+		resp := Response{
+			Context: Context{
+				Error: Error{
+					ErrorType: InputError,
+					Message:   "Invalid Input",
+				},
+			},
+		}
+		out, _ := json.Marshal(resp)
+		fmt.Fprintf(w, string(out))
+		return
+	}
+
+	resp, _ := router.Delegate(body)
+
+	out, _ := json.Marshal(resp)
+
+	fmt.Fprintf(w, string(out))
+}
+
+func main() {
+	numServers, err := strconv.Atoi(os.Getenv(serversEnvVar))
+	if err != nil || numServers < 1 {
+		numServers = 1
+	}
+
+	serverCmd := os.Getenv(serverCmdEnvVar)
+	servers, err = createServers(numServers, serverCmd)
+	router = NewRouter(servers)
```

Comment:
seems like `servers` could be kept inside `Router`,  and Router could implement `Shutdown` method to shut down the servers. `Shutdown` could be then called from signal handler. This would remove the need for global variable.